chunked_stream/lib/src/chunked_stream_iterator.dart

```diff
@@ -23,197 +23,179 @@ import 'dart:math';
 /// `List<int>` for each event.
 ///
 /// Note. methods on this class may not be called concurrently.
-class ChunkedStreamIterator<T> {
-  StreamSubscription<List<T>> _subscription;
-  bool _isDone = false;
-  List<T> _lastChunk;
-  Object _error;
-  StackTrace _stackTrace;
-  void Function() _wakeup;
-
-  List<T> _buffered = [];
-
-  ChunkedStreamIterator._(Stream<List<T>> stream) {
-    ArgumentError.checkNotNull(stream, 'stream');
-
-    _subscription = stream.listen((data) {
-      _lastChunk = data;
-      _wakeup();
-    }, onDone: () {
-      _isDone = true;
-      _wakeup();
-    }, onError: (e, st) {
-      _error = e;
-      _stackTrace = st;
-      _wakeup();
-    });
-    _subscription.pause();
+abstract class ChunkedStreamIterator<T> {
+  factory ChunkedStreamIterator(Stream<List<T>> stream) {
+    return _ChunkedStreamIterator<T>(stream);
   }
 
-  /// Create a [ChunkedStreamIterator] over [stream].
-  factory ChunkedStreamIterator(Stream<List<T>> stream) =>
-      ChunkedStreamIterator._(stream);
+  /// Returns a list of the next [size] elements in original chunks. The size of
+  /// the chunks are not guaranteed to be consistent.
+  ///
+  /// Returns a list with a total of less than [size] elements split in chunks
+  /// if the end of stream is encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  Future<List<List<T>>> readPreservingChunks(int size);
 
-  void _reset() {
-    _wakeup = null;
-    if (!_subscription.isPaused) {
-      _subscription.pause();
-    }
-  }
+  /// Returns a list of the next [size] elements.
+  ///
+  /// Returns a list with less than [size] elements if the end of stream is
+  /// encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  Future<List<T>> read(int size);
 
-  /// Read a series of chunks from the stream until we have [size] number of
-  /// items, then return these items.
+  /// Cancels the stream iterator (and the underlying stream subscription)
+  /// early.
   ///
-  /// This returns less than [size], if end of stream occurs before [size]
-  /// number of items have been received.
+  /// The [ChunkedStreamIterator] is automatically cancelled if [read] or
+  /// [readPreservingChunks] reaches the end of the stream or an error.
```

Comment:
What if `read` or `readPreservingChunks` throws an exception? Also why is this not documented in `read` or `readPreservingChunks`?

I think these semantics are fine.. I'm not sure it's normal or expected.. Honestly, I have a hard time grasping what is right. In most case this is probably preferable. We could add a boolean to the constructor `cancelOnError: true`.

Also what if an error happens while reading a `substream`? I almost expect the same behavior, I'm guessing it's just not documented.